File: engine/models/formation.ts
import { Position } from "@/engine/enums";

/** A formation defines the 11 positions for the starting lineup. */
export interface Formation {
  name: string;
  positions: readonly Position[];
}

export const F442: Formation = {
  name: "4-4-2",
  positions: [
    Position.GK,
    Position.LB,
    Position.CB,
    Position.CB,
    Position.RB,
    Position.LM,
    Position.CM,
    Position.CM,
    Position.RM,
    Position.ST,
    Position.ST,
  ],
} as const;

export const F433: Formation = {
  name: "4-3-3",
  positions: [
    Position.GK,
    Position.LB,
    Position.CB,
    Position.CB,
    Position.RB,
    Position.CM,
    Position.CM,
    Position.CAM,
    Position.LW,
    Position.ST,
    Position.RW,
  ],
} as const;

export const F352: Formation = {
  name: "3-5-2",
  positions: [
    Position.GK,
    Position.CB,
    Position.CB,
    Position.CB,
    Position.LM,
    Position.CDM,
    Position.CM,
    Position.CAM,
    Position.RM,
    Position.ST,
    Position.ST,
  ],
} as const;

export const F4231: Formation = {
  name: "4-2-3-1",
  positions: [
    Position.GK,
    Position.LB,
    Position.CB,
    Position.CB,
    Position.RB,
    Position.CDM,
    Position.CDM,
    Position.LW,
    Position.CAM,
    Position.RW,
    Position.ST,
  ],
} as const;

export const F532: Formation = {
  name: "5-3-2",
  positions: [
    Position.GK,
    Position.LWB,
    Position.CB,
    Position.CB,
    Position.CB,
    Position.RWB,
    Position.CM,
    Position.CM,
    Position.CAM,
    Position.ST,
    Position.ST,
  ],
} as const;

export const FORMATION_PRESETS: readonly Formation[] = [
  F442,
  F433,
  F352,
  F4231,
  F532,
];

File: src/data/Models/Formation.cs
using ElevenLegends.Data.Enums;

namespace ElevenLegends.Data.Models;

/// <summary>
/// A formation defines the 11 positions for the starting lineup.
/// </summary>
public sealed record Formation
{
    public required string Name { get; init; }
    public required IReadOnlyList<Position> Positions { get; init; }

    public static readonly Formation F442 = new()
    {
        Name = "4-4-2",
        Positions = [Position.GK, Position.LB, Position.CB, Position.CB, Position.RB,
                     Position.LM, Position.CM, Position.CM, Position.RM,
                     Position.ST, Position.ST]
    };

    public static readonly Formation F433 = new()
    {
        Name = "4-3-3",
        Positions = [Position.GK, Position.LB, Position.CB, Position.CB, Position.RB,
                     Position.CM, Position.CM, Position.CAM,
                     Position.LW, Position.ST, Position.RW]
    };

    public static readonly Formation F352 = new()
    {
        Name = "3-5-2",
        Positions = [Position.GK, Position.CB, Position.CB, Position.CB,
                     Position.LM, Position.CDM, Position.CM, Position.CAM, Position.RM,
                     Position.ST, Position.ST]
    };

    public static readonly Formation F4231 = new()
    {
        Name = "4-2-3-1",
        Positions = [Position.GK, Position.LB, Position.CB, Position.CB, Position.RB,
                     Position.CDM, Position.CDM,
                     Position.LW, Position.CAM, Position.RW,
                     Position.ST]
    };

    public static readonly Formation F532 = new()
    {
        Name = "5-3-2",
        Positions = [Position.GK, Position.LWB, Position.CB, Position.CB, Position.CB, Position.RWB,
                     Position.CM, Position.CM, Position.CAM,
                     Position.ST, Position.ST]
    };

    public static readonly IReadOnlyList<Formation> Presets = [F442, F433, F352, F4231, F532];
}
